sensors/baro/ms5611.c

```diff
@@ -0,0 +1,280 @@
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/time.h>
+#include <sys/threads.h>
+#include <spi.h>
+#include <sensors-spi.h>
+
+#include "../sensors.h"
+
+/* MS5611 COMMANDS */
+
+/* Reset command */
+#define CMD_RESET 0x1e
+
+/* Conversion request command and oversampling (OSR) rate flags */
+#define CMD_CVRT_PRESS 0x40
+#define CMD_CVRT_TEMP  0x50
+#define CVRT_OSR_256   0x00
+#define CVRT_OSR_512   0x02
+#define CVRT_OSR_1024  0x04
+#define CVRT_OSR_2048  0x06
+#define CVRT_OSR_4096  0x08
+
+/* Read ADC conversion result command */
+#define CMD_READ_ADC 0x00
+
+/* Read PROM command and PROM address specifiers */
+#define CMD_READ_PROM 0xa0
+#define PROM_SIZE     8
+
+/* MS5611 DELAYS */
+
+/* OSR dependent minimum wait (microseconds) for conversion. Datasheet values ceil-ed to full milliseconds */
+#define OSR_256_SLEEP  1000
+#define OSR_512_SLEEP  2000
+#define OSR_1024_SLEEP 3000
+#define OSR_2048_SLEEP 5000
+#define OSR_4096_SLEEP 10000
+
+/* Device reset duration */
+#define RESET_SLEEP 3000 /* Datasheet says 2800. It seems some data may be lost with such sleep */
+
+
+typedef struct {
+	spimsg_ctx_t spiCtx;
+	oid_t spiSS;
+	sensor_event_t evtBaro;
+	uint16_t prom[PROM_SIZE]; /* Prom memory: [ reserved, C1, C2, C3, C4, C5, C6, CRC ] */
+	char stack[512] __attribute__((aligned(8)));
+} ms5611_ctx_t;
+
+
+/* Based on sample code in AN520 application note for MS56xx/MS57xx/MS58xx pressure sensor by Measurement Specialties */
+uint8_t ms5611_crc4(const uint16_t *prom)
+{
+	uint16_t cnt, bit, word;
+	uint32_t rem = 0; /* crc reminder */
+
+	/* Operation is performed on bytes */
+	for (cnt = 0; cnt < (PROM_SIZE * 2); cnt++) {
+		/* The CRC code is calculated and written in factory with the LSB byte in the prom (prom[7]) set to 0x00 */
+		word = (cnt == (PROM_SIZE * 2 - 1)) ? 0 : prom[cnt >> 1];
+
+		/* High/low byte selection */
+		if (cnt & 1 == 1) {
+			word <<= 8;
+		}
+		rem ^= word >> 8;
+
+		for (bit = 8; bit > 0; bit--) {
+			rem = ((rem & 0x8000) > 0) ? (rem << 1) ^ 0x3000 : (rem << 1);
+		}
+	}
+
+	rem = (0x000f & (rem >> 12)); /* final 4-bit reminder is CRC code */
+
+	return (rem ^ 0x00);
+}
+
+
+/* Sends conversion request `convCmd`, waits `usDelat` for conversion and reads ADC to `res` */
+static int ms5611_measure(ms5611_ctx_t *ctx, uint8_t convCmd, time_t usDelay, uint32_t *res)
+{
+	static const uint8_t readCmd = CMD_READ_ADC;
+	uint8_t val[3];
+
+	if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &convCmd, sizeof(convCmd), NULL, 0, 0) < 0) {
+		fprintf(stderr, "ms5611: failed conv. request\n");
+		return -1;
+	}
+
+	usleep(usDelay);
+
+	if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &readCmd, sizeof(readCmd), val, sizeof(val), sizeof(readCmd)) < 0) {
+		fprintf(stderr, "ms5611: adc read failed\n");
+		return -1;
+	}
+
+	/* If ADC is read too early then it returns 0 */
+	if (val[0] == 0 && val[1] == 0 && val[2] == 0) {
```

Comment:
This check would be more efficient to perform on `res`, as this needs 3 compares instead of 1 for uint32_t.